netket/experimental/driver/vmc_srt.py

```diff
@@ -0,0 +1,202 @@
+        jacobian_mode: Optional[str] = None,
+        variational_state: VariationalState = None,
+    ):
+        """
+        Initializes the driver class.
+
+        Args:
+            hamiltonian: The Hamiltonian of the system.
+            optimizer: Determines how optimization steps are performed given the
+                    bare energy gradient.
+            diag_shift: The diagonal shift of the stochastic reconfiguration matrix.
+                        Typical values are 1e-4 ÷ 1e-3. Can also be an optax schedule.
+            hamiltonian: The Hamiltonian of the system.
+            linear_solver_fn: Callable to solve the linear problem associated to the
+                              updates of the parameters
+            jacobian_mode: The mode used to compute the jacobian of the variational state. Can be `'real'`
+                    or `'complex'` (defaults to the dtype of the output of the model).
+            variational_state: The :class:`netket.vqs.MCState` to be optimised.
+        """
+        super().__init__(hamiltonian, optimizer, variational_state=variational_state)
+
+        self._ham = hamiltonian.collect()  # type: AbstractOperator
+        _, self.unravel_params_fn = ravel_pytree(self.state.parameters)
+        self.diag_shift = diag_shift
+        self.jacobian_mode = jacobian_mode
+        self._linear_solver_fn = linear_solver_fn
+
+    @property
+    def jacobian_mode(self) -> str:
+        """
+        The mode used to compute the jacobian of the variational state. Can be `'real'`
+        or `'complex'`.
+
+        Real mode truncates imaginary part of the wavefunction, while `complex` does not.
+        This internally uses :func:`netket.jax.jacobian`. See that function for a more
+        complete documentation.
+        """
+        return self._jacobian_mode
+
+    @jacobian_mode.setter
+    def jacobian_mode(self, mode: Optional[str]):
+        if mode is None:
+            mode = nkjax.jacobian_default_mode(
+                self.state._apply_fun,
+                self.state.parameters,
+                self.state.model_state,
+                self.state.samples,
+                warn=False,
+            )
+
+        if mode not in ["complex", "real"]:
+            raise ValueError(
+                "`jacobian_mode` only supports 'real' for real-valued wavefunctions and"
+                "'complex'.\n\n"
+                "`holomorphic` is not yet supported, but could be contributed in the future."
+            )
+        self._jacobian_mode = mode
+
+    def _forward_and_backward(self):
+        """
+        Performs a number of VMC optimization steps.
+
+        Args:
+            n_steps (int): Number of steps to perform.
+        """
+
+        self.state.reset()
+
+        # Compute the local energy estimator and average Energy
+        local_energy = self.state.local_estimators(self._ham).squeeze()
+
+        e_mean = local_energy.mean()
```

Comment:
@riccardo-rende Are you sure here it's a mean, and not a mean over MPI as well?

Reply:
I think it was wrong. I changed it to an mpi average. it is passing tests under MPI now.

Reply:
please confirm

Reply:
Yes, it was wrong! I was also able to test the new version using MPI and it works! Thank you very much.